flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java

```diff
@@ -230,7 +227,6 @@ private BufferBuilder getBufferBuilder(int targetChannel) throws IOException, In
 	}
 
 	private BufferBuilder requestNewBufferBuilder(int targetChannel) throws IOException, InterruptedException {
-		checkState(!bufferBuilders[targetChannel].isPresent());
```

Comment:
This `checkState` was preventing from some bugs and data loses. Maybe replace with:
`checkState(!bufferBuilders[targetChannel].isPresent() || bufferBuilders[targetChannel].isFinished());`?

Reply:
The calling from `getBufferBuilder` does not need this `checkState`. But for calling from `copyFromSerializerToTargetChannel`, it may be necessary to add this check for avoiding bugs.

I removed it just for reducing some overheads. I will restore this check. :)